### ISSUE TITLE
Precedences for prefix operators

### ISSUE THREAD

**Issue by 01mf02** [Deducteam/lambdapi#244]
It would be very convenient for Lambdapi to support custom precedences for prefix operators, similar to the existing support for infix operators. Note that it is possible to define prefix operators, as follows:

~~~
definition not : eta (o > o) ≔ λ p, p --> false
set declared "~"
definition ~ : eta (o > o) ≔ not
~~~

However, the operator `~` then always binds just like any other function symbol, such as `not`. This really shows in the following example:

~~~
// these expressions typecheck
type λ p, ~ p
type λ p, ~ (~ p)
type λ p x, ~ (p x)

// these expressions do not typecheck, but I would like them to!
type λ p, ~ ~ p
type λ p x, ~ p x
~~~

Such expressions regularly appear in Coq and Isabelle. I ported some Coq proofs to Lambdapi, and forgetting to put parentheses around negated expressions was by far my most frequent error (this alone probably cost me hours).

I would like to have something like:

~~~
definition not : eta (o > o) ≔ λ p, p --> false
set prefix 40 "~" ≔ not
~~~

Similar syntax in Isabelle:

~~~
definition not :: "o ⇒ o"  ("¬ _" [40] 40)
  where "not ≡ λA. A ⟶ False"
~~~

**Comment by 01mf02** [comment]
A full example follows:

~~~
symbol const Type : TYPE
symbol injective eta : Type ⇒ TYPE

// function type
symbol const Ar : Type ⇒ Type ⇒ Type
set infix right 6 ">" ≔ Ar
rule eta (&a > &b) → eta &a ⇒ eta &b

symbol const i : Type
symbol const o : Type

symbol eps : eta o ⇒ TYPE

symbol const imp : eta (o > o > o)
rule eps (imp &a &b) → eps &a ⇒ eps &b
set infix right 6 "-->" ≔ imp

symbol const all : ∀ {A : Type}, eta ((A > o) > o)
rule eps (@all &a &b) → ∀ (x : eta &a), eps (&b x)

definition false : eta o ≔ all (λ p, p)

definition not : eta (o > o) ≔ λ p, p --> false
set declared "~"
definition ~ : eta (o > o) ≔ not

// these expressions typecheck
type λ p, ~ p
type λ p, ~ (~ p)
type λ p x, ~ (p x)

// these expressions do not typecheck, but I would like them to!
type λ p, ~ ~ p
type λ p x, ~ p x
~~~

**Comment by rlepigre** [comment]
I implemented prefix operators in a PR, let me know what you think. Also I was not able to make your examples `λ p x, ~ (p x)` and `λ p x, ~ p x` type-check (although they seem to be parsed correctly). I might be missing something here since I have pretty bad intuitions with encodings.

**Comment by fblanqui** [comment]
The problem with ~(p x) and ~ p x is related to unification. I will investigate this.

**Comment by fblanqui** [comment]
Unification problem for ~(p x) fixed in 3b845c528e723bf992ebca6f68290d61c107cb8d.
